Create citation from crossref

LukasWallrich opened the pull request:
Thanks for this package! In line with #2, I would love to create citations directly from DOIs, using crossref data. I have made a first stab at implementing this here - hope it might be helpful. 

Of course, I'd be happy to make any edits you suggest - instead of it being a separate function, for instance, drop_name could also check whether bib consists of DOIs and use crossref automatically in that case?

nucleic-acid commented:
Hi, sorry for not getting back earlier. Thanks for this great contribution! As you've probably seen this is one of the earliest extensions that I wanted to implement, so the contribution is very much appreciated!

I'll have time looking at the code later but in general I prefer your approach to have a separate wrapper function since drop_name is already quite bloated.

As I don't have any experience with the rcrossref package, I was wondering what kind of identifiers you can use? I had in mind to also provide the possibility to provide PMIDs (biomedical field) and maybe arXiv IDs.
Do you know if that is possible via crossref?

nucleic-acid commented on R/drop_name_crossref.R:
```diff
@@ -0,0 +1,46 @@
+    authors <- dplyr::pull(dplyr::mutate(x$author, name = paste0(family, ", ", given)))
+    author_last <- x$author$family[1]
+    dplyr::tibble(YEAR = x$created$`date-parts`[1], JOURNAL = x$`container-title`, AUTHOR = list(authors),
+                  TITLE = x$title, BIBTEXKEY = paste(author_last, x$created$`date-parts`[1]), DOI = x$DOI)
```
I would prefer either a paste0 at this point or specifying a separator like _ to prevent filenames with spaces.

LukasWallrich replied:
Makes sense. However, `author_last` will often contain spaces as well - so should we rather wrap the whole key into stringr::str_replace(___, " ", "_")?

nucleic-acid replied:
that sounds good.

nucleic-acid commented on R/drop_name_crossref.R:
```diff
@@ -0,0 +1,46 @@
+
+  stopifnot(is.character(dois))
+
+  ref <-  rcrossref::cr_cn(dois, "citeproc-json")
```
What would you say, if we wrapped this into a try-catch environment? If this fails, we could provide solutions to common issues, such as not providing one's mail address to cross-ref, etc..

LukasWallrich replied:
Yes - even though my understanding is that providing the email address is optional. But given the issues you experienced that seems necessary.

nucleic-acid replied:
I haven't used the rcrossref package before, so Im not too deep into the details. I came across [one stackoverflow answer](https://stackoverflow.com/questions/68412035/rcrossref-api-error-resresponse-headerscontent-type-application-jsonch) that made it seem it's mandatory nowadays. (The linked github issue doesn't say that, though). Better safe than sorry.

nucleic-acid commented on R/drop_name_crossref.R:
```diff
@@ -0,0 +1,46 @@
+#'
+#' @param dois One or several dois to create visual citations for. If they are named, these names are used as filenames; otherwise they are generated based on authors and years.
+#' @inheritDotParams drop_name -bib -cite_key
+#'
```
drop_name_crossref should behave like drop_name and return a file path to keep things consistent. I would add this to the documentation, like e.g.

`#' @return A character string with the file path to the created visual citation in the specified output format.`

nucleic-acid commented:
This is a concise and straight forward implementation, thanks for the work! I also learned a lot by reading your code.

I have a few minor points to debate and one general problem, that in my view prevents inclusion of this PR in the next release at the moment.

## minor comments
See comments above for the minor points.

## major concern
The major issue is not related to the proposed code, but to the {rcrossref} package:
I could not access the crossref API through the CRAN version of {rcrossref}, even after adding my mail address to Renviron. According to [this issue from Feb 2022](https://github.com/ropensci/rcrossref/issues/227) you need the development version of {rcrossref} to obtain data through the API. I want to keep the package operable for people who cannot (or don't want to) install dev versions.

Do you have an idea how to solve this issue? I'm always open for suggestions, maybe I missed some info or did not use the rcrossref package correctly?

My suggestion _for the time beeing_, until the latest API-calls are accessible in the CRAN version of {rcrossref} would be to put the function you wrote into a new vignette. This way, people who want to use DOIs directly can implement the function locally together with the dev version of {rcrossref} by just copy-pasting the code from the vignette, while the {namedropR} package would not rely on dev-version-dependencies.

LukasWallrich commented:
Thanks for your positive feedback - as indicated, I'm happy to address the minor points. Re the major point, I have just asked in the rcrossref repo when they are planning to submit the next version to CRAN ... if that happens soon that might be the easiest solution, otherwise the vignette idea sounds good as well.

Re other identifiers: rcrossref will only resolve DOIs (but usually find almost all of them) ...

LukasWallrich commented:
I just added the edits you proposed, as well as a check whether any citations were returned (given that rcrossref only gives a warning rather than an error when a DOI is not found).

Apparently, the rcrossref CRAN package will be updated shortly (within the next two weeks) so I think we might want to wait for that?

I also added an example to the documentation - do you think it should be documented any further? In the README?
